Seven Billion Hour Wait Time - Or, Expanded Job Config (TOML time)

san7890 opened the pull request:
a TOML is just a sexier JSON (bullet flies through my head, killing me instantly)

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Hey there!

Twenty-eight days ago, there was a PR up (#69648, which this is an alternative to) that would allow you to define the Minimum Required Playtime (in minutes) or the Minimum Account Age (in days) to unlock a job in config! That was great, but it unfortunately caused too many changes to an existing system, and it staled out on the workbench before it got anywhere. I don't think it's fair to have Cyborgs wait 21 days to Cyborg, especially when it's not that high-impact as a role, so I wrote up a new system.

We're going to use TOML for this new system. This isn't solely because it's new and fancy and I wanted to try out something: there's a few discrete benefits that I wanted to harness.

* Extensibility: In case we want to add even more variables pertaining to jobs in config, we don't have to write up even more contrived RegEx bullshit in order to parse a system, we can just define a string in the TOML format.

* Readability: It's much more readable for me to clearly see what value pertains to what value rather than have to memorize A,B,C,D... etc.

* Verifiability: With this, you should be able to better parse when something's invalid or mis-matches between the code and the config, which isn't something I think would be possible in the old text format. Let me know if that's incorrect.

Anyways, technical considerations. The old jobs.txt format is still supported codeside, and I wrote it such that the TOML file not existing would pass a small message, but still load the old jobs.txt. I think that's what you guys wanted after all that conversation? Sorry if I fucked that up.

Anyways, that should be all? Everything in jobconfig.toml is codebase default values (passed on this PR). Keep in mind that neither jobs.txt nor jobconfig.toml will be used if the LOAD_JOBS_FROM_TXT flag is never enabled in config. Hope that helps.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

it breaks my heart every time someone adminhelps asking why they can't play cyborg for another 18 days and all i can do is shake a fist and grumble "coders..." as a tear streaks down my face.

---
The old system working on the failover code (jobs.txt exists, using that instead of jobconfig.json). 50 Captains sounds like a great idea for a proud new Headmin term, doesn't it?:

![image](https://user-images.githubusercontent.com/34697715/192919005-81885849-c0fe-4462-bfbe-bc7a55c99740.png)

The new TOML loading system working (I lowered the captain's required account age by two days, aren't I a goober?):

![image](https://user-images.githubusercontent.com/34697715/193703852-6c83c589-ebab-4641-b69a-f69141fb7ff4.png)

---

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: san7890, with regards to FernandoJ8
config: A new TOML system has been introduced to modify different settings and variables regarding jobs! You can edit the standard stuff, like roundstart positions and overall positions of a job like usual (feature parse), but you can edit two new things! You can change how long an account needs to be "aged" on a server for to unlock a job, as well as the actual number of minutes a player must get to unlock a specific job. Nifty, eh? Do remember, this is entire system won't load without the overall LOAD_JOBS_FROM_TXT flag!

server: Keep in mind that the jobconfig.toml is now the preferred system for updating job-related config, while jobs.txt should be considered deprecated. Support for loading it via jobs.txt will remain indefinitely, but expect to see a small error every time we validate the config.

server: FYI: To use jobconfig.toml, it is strongly advised that instead of downloading the codebase default that you run the Generate Job Configuration verb (in Server tab) in order to import all of your already set variables from jobs.txt. Please check to ensure that all of your wanted settings are present in the new file.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Again, the system of deprecating a config in favor for a new system is based on snippets I've heard of MSO talking on the subject and something mothblocks said once. Let me know if anything is absurdly cringe and I'll change it (do tell me specifically what you want though).

san7890 committed:
Seven Billion Hour Wait Time - Or, Expanded Job Config (JSON time)

=Hey there!

Twenty-eight days ago, there was a PR up that would allow you to define the Minimum Required Playtime (in minutes) or the Minimum Account Age (in days) to unlock a job in config! That was great, but it unfortunately caused too many changes to an existing system, and it staled out on the workbench before it got anywhere. I don't think it's fair to have Cyborgs wait 21 days to Cyborg, especially when it's not that high-impact as a role, so I wrote up a new system.

We're going to use JSON for this new system. This isn't solely because it's new and fancy and I wanted to try out something: there's a few discrete benefits that I wanted to harness.

* Extensibility: In case we want to add even more variables pertaining to jobs in config, we don't have to write up even more contrived RegEx bullshit in order to parse a system, we can just define a string in the JSON format.

* Readability: It's much more readable for me to clearly see what value pertains to what value rather than have to memorize A,B,C,D... etc.

* Verifiability: With this, you should be able to better parse when something's invalid or mis-matches between the code and the config, which isn't something I think would be possible in the old text format. Let me know if that's incorrect.

Anyways, technical considerations. The old jobs.txt format is still supported codeside, and I wrote it such that the txt format would always overwrite the JSON format if it existed because that made sense in my head. Let me know if you have a better idea, I'm just basing this off a comment MSO left in the aforementioned PR. I wrote a warning to the codebase side jobs.txt that new server operators should delete it and migrate to jobconfig.json, let me know if that is cringe. I also took a leaflet out of (MSO's?) book in regards to having an error post when we hit a major BYOND version 2 ahead of our present (514 -> 516) because server operators should have switched over by then so we can update the code. Again, lemme know if that's cringe (but I think it's rather intuitive).

Anyways, that should be all? Everything in jobconfig.json is codebase default values. Keep in mind that neither jobs.txt nor jobconfig.json will be used if the LOAD_JOBS_FROM_TXT flag is never enabled in config. Hope that helps.

san7890 committed:
updates a comment to be actually correct

Mothblocks commented:
your indentation is whacked up you should use toml because it has every benefit you have except its flattened and nicer to read and a language actually designed to be a config (and its usually very good at it, other than when things get VERY deep)

```toml
["Captain"]
"Total Positions" = 1
"Spawn Positions" = 1

["Head of Personnel"]
"Etc" = 1
"Etcccc" = 1
```

Mothblocks commented:
i dont really think we need the DM_VERSION thing either. its more enticing for something that isnt so trivial to keep backwards compatibility with

Mothblocks commented:
if you want to go above and beyond you could automatically migrate the existing config over to a new one just by creating the file btw

san7890 commented:
> if you want to go above and beyond you could automatically migrate the existing config over to a new one just by creating the file btw

Do you mean if the `jobconfig.json` (as of the current state of this PR) fails, we create `config/jobconfig.json` with the default codebase values by reading out of `jobs.txt` in some manner?

Mothblocks commented:
yes

YakumoChen commented:
> I don't think it's fair to have Cyborgs wait 21 days to Cyborg

I'd say a role with all access and the requirement to follow a specific set of rules, to the letter, that can change on a whim is more of a skill floor role than high impact, hence the timer.

Mothblocks commented:
That's for policy discussion to decide now

optimumtact commented:
I'm not very keen on json as a configuration language and I don't see what was missing in our existing config parser that you needed json for.

Mothblocks commented:
im not sure if youre confusing toml with ini but ini is explicitly not specified/standard while toml is (you could call toml a standardized ini and be mostly right)

Mothblocks commented:
or if by ini format you mean the old jobs.txt actually

optimumtact commented:
All of our config files follow the basic ini format of 
NAME = expression

san7890 commented:
I wanted something more expandable than a text file that demands you sit at a RegEx parser for 15 minutes (assuming someone at my skill level) to get some deeper nuance

optimumtact commented:
>I wanted something more expandable than a text file that demands you sit at a RegEx parser for 15 minutes (assuming someone at my skill level) to get some deeper nuance.

dude what the hell are you talking about?

san7890 commented:
> > I wanted something more expandable than a text file that demands you sit at a RegEx parser for 15 minutes (assuming someone at my skill level) to get some deeper nuance.
> 
> dude what the hell are you talking about? What functionality were you trying to achieve that you could not achieve

I like not having to memorize which group goes where when I have to edit something, like I would rather have each number be associated to a key every single time rather than have to try and remember a pattern of what I need to edit. This is trivial for jobs.txt in it's current form (which only consists of two values), but we're growing it now to double that amount, and I want to leave the door open for future extensibility.

Other than having a longer file, I think the enhanced clarity this provides is a benefit.

optimumtact commented:
San, for the love of god, please EXPLAIN, why is it 3 comments deep into this conversation and you haven't managed to actually form a comprehendable reason why or example that you need something more than the basic config parsing.

optimumtact commented:
We spoke about this in voice, the reason is that san want's the capability of grouping a set of keys under a single category group key (i.e individual config options for a job)

san7890 committed:
auto-generation of the JSON and stuff

renamed the proc because it was stupid and sucked because it was just synonymous to another proc and was not distinct what-so-ever

This new system will delete the jobs.txt and create a new jobconfig.json. Tell me if that is cringe.

san7890 commented:
@Mothblocks I got 2/3:

Removed the BYOND 516 check because we now migrate from the TXT to the JSON
Autogeneration of the JSON file based on the old TXT.

There is no way to create (like a toml_encode proc) a TOML, so that stinks. The auto-generated JSON is now ugly as hell, but it works:tm: and you can JSON prettify it on the box later on. I couldn't find anything to help prettify the JSON, let me know if there is a way to do that.

I also have it fdel the jobs.txt when you migrate, let me know if that's cringe too and I can fix that.

Keeping this drafted since we want to have a specific config_key tied to the job datum that we track instead of tying it to the name of the job, so I'm going to work on that later.

optimumtact commented:
>Keeping this drafted since we want to have a specific config_key tied to the job datum that we track instead of tying it to the name of the job, so I'm going to work on that later.

This was my request so we can hopefully migrate away from the display name of the job also being it's key later on.

san7890 committed:
whoops, fucked up a few things

patches up some inconsistencies before i embark on more project stuff

san7890 committed:
comments r good

san7890 committed:
that was an inaccurate comment

Mothblocks commented:
> I could not find a way to create (like a toml_encode proc) a TOML

Oh that's a sad oversight

Ok I'll add that to rust-g

san7890 commented:
> > I could not find a way to create (like a toml_encode proc) a TOML
> 
> Oh that's a sad oversight
> 
> Ok I'll add that to rust-g

yeah i could definitely add it in this PR and bump rust-g if you can get it in

Mothblocks commented:
Just prefix it with the comment that's what everyone does

Mothblocks commented:
```dm
var/code = rustg_toml_encode(whatever)
rustg_write_file("# This file was @autogenerated from automatic migration\n[code]", path)
```

Mothblocks commented:
https://github.com/tgstation/rust-g/pull/116

Will get this in and a new rust-g build in a few days tops

san7890 committed:
Wanted Updates In Prep For TOML Time

* writes a proc to rapidly "regenerate" the config, very nice

* writes out the lists

* config key time whoopie yippie wahoo

san7890 committed:
i pushed to remote too early

san7890 committed:
i fucked up and went goober mode

san7890 committed:
re-makes jobs.txt for testing purposes

san7890 committed:
immediately kills jobs.txt again so i can test migration later

i'm doing this so i can easily revert one commit and test the migration to ensure it'll work when it's TOML time

san7890 committed:
and a few dollars more

MrStonedOne commented:
From a quick glance this entire system seems like it has the issue that it doesn't properly seperate default values from configured values that happen to be the same as the default value and that will byte us in the ass when we go to update or change any of the code default values.

This should function like a set of overrides from the code default such that we can change the code defaults and have them apply to any server or downstream that did care to change them.

**For this reason, it can not auto-create or auto-convert. the file not existing has to be a valid state.**

Also i should point out that i will not support lowing of cyborg or ai age restrictions. This system existing was basically the only reason I as a player cared to follow the rules. I spent 21 days waiting for round start cyborg and 30 days waiting for round start AI, and i wasn't gonna fucking do that again. (As a player I started from the assumption that I could successfully ban evade, and despite not knowing what I know about cids, I likely could have if I spent enough time on it.)

The account age system and the job exp system being two seperate systems that run concurrently is intentional. One creates value in keeping an account in good standing to avoid rule breakers, one keeps inexperienced people out of roles they aren't ready for.

Mothblocks commented:
> For this reason, it can not auto-create or auto-convert. the file not existing has to be a valid state.

these two things arent opposed to each other though. i just mean migrating over the existing config over to the new config?

MrStonedOne commented:
the code san currently has will create the new file with code defaults if neither exist.

Mothblocks commented:
makes sense. definitely should be fixed

MrStonedOne commented:
and honestly, given that i don't think we even use the jobs config much for anything, i don't see the need to autoconvert.

1: game code that writes to and deletes files should be minimized, especially code that deletes files. more so given this is server operator managed files in `config/`, not game managed files (`data/`)

2: usage of this entire config system is likely very low. My guess is most people will end up making a fresh brand new file for this purely to change some of the time requirements.

oh, san, one idea, fuck toml, it doesn't even let you do nested/hierarchy shit properly

sync up with me, we should work on porting the map config driver to the config system as a config type, it would be perfect for this.


```
job "Clown"
	spawn_positions 420
	max_positions 666
	account_age 69
endjob
```

Mothblocks commented:
doesnt san want to be able to easily extract the data in another language without writing a parser or whatever oranges was trying to squeeze out of him

san7890 commented:
Yeah, the whole purpose of this PR was to step away from the need to have a text-based parser written out somewhere, and although I do understand it's merits, my understanding of TOML hierarchies is that it gets messy on something that is more-than-trivial. I don't know if it's possible to scale this system with such increasing key complexity that you would run into those issues (although it's not fair to never-say-never).

I'm quite happy with how the TOML file is working out for what I am able to comprehend looking at it at from at both a developer's viewpoint as well as coming from the viewpoint of someone who has to end up working with these systems at the end-user's point. Although it may not be that way at first, I know that if I (specifically me) were to migrate into a text-based hierachy system like map config loader, it would utilize a very barebones parser with minimal modifications and go right back to that issue of being crippled by in-extensibility rather than just be able to whip up a new nested key that can rest under the job. Please do let me know if my personal philosophy on this is directly against the stream of your desires and I can update that accordingly, but the initial tests with rustg_toml_encode have been promising from my development viewpoint.

Regardless, thank you for getting back to me on the "delete config file?" question. The reason I was concerned about it is because we both have `admin_nicknames.txt` and `admin_nicknames.json` on the server boxes (even though the code will only ever parse for `admin_nicknames.json`), and I wanted to see if there was a way to avoid that potential negative. However, I did recently realize that `jobs.txt` might include important comments/justification, and if a server isn't backing that up and the code auto-magically deletes it, that would be _horrible_ UX.

I'll rework the "migration" system. I don't believe that writing a new config file is the worst thing in the world, but I might just downgrade it to a new system where you can generate the auto-magically generated TOML file in-game and then download it to a location on your client's machine, and then put that on the box rather than have it directly write to the config folder. I'll also sit and think about the best way to approach the case where the respective config file is not found, and respect that as well. Thank you all for your consideration and patience thus far, my thoughts have been really scattered these last few days (and will continue to be scattered for a while).

Zirok-BYOND commented:
>extensibility
If a job was not present in the txt(or misspelled or whatever) this has runtimed(rather than providing a clean error message or simply failing silently) for 7+ years now, and the downstreams I've looked at have not changed this. Nobody cares about this or has looked at it, which should tell you just how low usage this is. Don't overengineer this.

san7890 commented:
> If a job was not present in the txt(or misspelled or whatever) this has runtimed(rather than providing a clean error message or simply failing silently) for 7+ years now, and the downstreams I've looked at have not changed this. Nobody cares about this or has looked at it, which should tell you just how low usage this is. Don't overengineer this.

Maybe this is less of the fault of the server operator and more of a fault of the prior existing system. I know that this can be solved during a rework of that system (which I will intend to do).

san7890 committed:
TOML System Mega-Commit

* Implements the TOML System (warning, uses unreleased rust-g.dll. I'll update this in a later commit and bump dependencies.sh)

* Implements jobconfig.toml! Also includes validation for this file on SSconfiguration's initialize, very nifty.

* Code will now respect the non-existence of jobconfig.toml and will no longer write to the `config/` directory at all. Instead, we will send an FTP prompt when someone runs the verb to generate the TOML file (because editing this shit manually sucks fucking cock and balls), and that'll still migrate from older versions of jobconfig.toml / jobs.txt

* Adds admin verb to generate the TOML file. Please let me know if you want it in a different category.

* That should be it? Let me know if there's some critical feedback that I missed out on that I should listen to. I'm starting to approach the end of passion for this project, and I don't know how much longer I can soldier though.

san7890 committed:
bumps rust-g.dll

the DLL added cleanly so i'm assuming that's good and just upping the version

san7890 commented:
Alright, just gonna paste the stuff I wrote for the last major commit here.

* Implements the TOML System and bumps rust-g.dll to leverage the new toml-encode system

* Implements jobconfig.toml! Also includes validation for this file on SSconfiguration's initialize, very nifty.

* Code will now respect the non-existence of jobconfig.toml and will no longer write or delete _at all_ to the `config/` directory at all. Instead, we will send an FTP prompt when someone runs the verb to generate the TOML file (because adding 35 new fields every time a coder sneezes sucks fucking cock and balls), and that'll still migrate from older versions of jobconfig.toml / jobs.txt

* Adds admin verb to generate the TOML file. Please let me know if you want it in a different category.

* That should be it? Let me know if there's some critical feedback that I missed out on that I should listen to. I'm starting to approach the end of passion for this project, and I don't know how much longer I can soldier though.

Let me know if there's a better way to do TOML validation since I am interested in accounting for edge cases in this PR though.

san7890 committed:
gets rid of caller

san7890 committed:
wait this is supposed to be commented out whoops

san7890 committed:
small consistency thing to ensure logs look good

san7890 committed:
okay kills caller for real

MrStonedOne commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +601,136 @@ SUBSYSTEM_DEF(job)
+	if(!legacy_mode) // this flag is set during the setup of SSconfig, and all warnings were handled there.
+		var/job_config = rustg_read_toml_file(toml_file)
+
+		for(var/datum/job/occupation as anything in joinable_occupations)
+			var/job_title = occupation.title
+			var/job_key = occupation.config_tag
+			if(!job_config[job_key])
+				message_admins(span_notice("[job_title] (with config key [job_key]) is missing from jobconfig.toml! Using codebase defaults.")) // we add both the title and the config key in case they desync over time and someone needs to fix the box for some reason
+				continue
```
Again, the job config has to work like an **override**.

If you want to stay subscribed to code defaults, you would do so by not including a job in the job toml file. if you only want to change some values from code default, you would do so by including only configuration options for those settings, and not including the values you don't want to change from default.

Ie, if i only want to change atmos to have a higher account age requirement from `0` days to `6` days, i should be able to just specify the following as the **entirety** of the toml and have this be valid:

```toml
[ATMOSPHERIC_TECHNICIAN]
"Playtime Requirements" = 60
```

Having to put every single job in the config and specify every single value, locking them in, is asking for trouble down the line when we want to be able to change the code defaults and have them take effect on the servers that didn't specify otherwise.

This also means that the entirety of the repo toml and the generated toml has to start commented out (except for anything specified in the legacy config, allowing people to uncomment things they want to change from repo default while serving as a documentation of what repo default is.

san7890 replied:
Okay, I think I understood it now. I'm pretty stupid and lazy (mostly, I couldn't think of a better way to do it), so I just purposefully introduced a key mis-match. Every time we generate jobconfig.toml, we will always write to it with commented out keys. However, if we're importing from jobs.txt and the variables were previously changed there, that will port as uncommented. If we regenerate and the specific line was uncommented, the behavior should allow that key to be uncommented in the downloadable file.

That's how it should work in theory. I'm not a very good programmer and I wanted to avoid bugs, so I probably did a bit more overengineering on this system that I needed to. I believe I got what you were saying here, though- and great apologies if I did not because I've had other things take up my time recently and I typically have to relearn what I was doing before working on the system again.

A briefer way of putting it: I'm not purposefully trying to irritate anyone. Thank you for your patience thus far.

san7890 committed:
BE WARE- DEMONS LAY HERE

you know that video with the animated stroke and it goes "wubuublubhbhbh (fart sound)". yeah.

san7890 commented:
as mentioned in the earlier code comments i am still novice so apologies if the patterns i used to make sure the intended behavior was achieved makes you want to throw up.

:(

Mothblocks commented:
relax

san7890 committed:
WAIT A SECOND THAT'S NOT RIGHT

* replaces old json decode with actual toml decode (loling, lmaoing).

* fixes some bad index bullshit i think this is the right way to fix it (using += to create assoc. lists)

* removes shitty indenting that accidentally broke shit very sad

* fixes paramedic config tag to actually work and not flag as mime (lol lmao).

san7890 committed:
small formatting addendum

san7890 committed:
paramedic aftermath

since the job keys were yorked the toml never generated correctly for reasons:tm: so let's actually pass a good one through  for the codebase default

Mothblocks commented on config/jobconfig.toml:
```diff
@@ -147,6 +147,12 @@
+"#Spawn Positions" = 1
+"#Total Positions" = 1
+
+[PARAMEDIC]
```
Is there a reason we're doing this instead of just `[Paramedic]`, `["Atmospheric Technician"]`, etc?

san7890 replied:
```dm
	/// String key to track any variables we want to tie to this job in config, so we can avoid using the job title. We CAPITALIZE it in order to ensure it's unique and resistant to trivial formatting changes.
	/// You'll probably break someone's config if you change this, so it's best to not to.
```

I wrote this in the DMDoc for `config_tag`, and I based it around the conversation I had with oranges when he was giving me a few examples on what he wanted the system to be resistant to, like if we changed it from `Station Engineer` to like `Station engineer` (this is the example he gave me), and I kinda wanted to emphasize that it was a key and not the actual title of the job by having that be `STATION_ENGINEER`. Let me know if that's bad UX or something.

Mothblocks replied:
Fair enough

san7890 committed:
small documentation update

it was kinda silly

san7890 committed:
just realized i was doing something in a crappy way

san7890 committed:
full ass export_toml proc

less copypasta is good i think

also some config documentation fix

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -781,7 +774,14 @@ SUBSYSTEM_DEF(job)
+	var/payload = rustg_toml_encode(data)
 	var/temp_file = file("data/jobconfig.toml")
 	if(fexists(temp_file))
 		fdel(temp_file) // ensure it writes properly in case it exists
```
Use `rustg_file_write` which will completely overwrite a file

san7890 committed:
uses rustg_file_write instead

yeah this is poggers

san7890 committed:
outdated documentation (very bad)

yeah this was true a week ago but it's not at all accurate right now

san7890 committed:
wait i cached it

why was i having it go through the file AGAIN uuuugh

san7890 committed:
i knew it fucking didn't work correclty fuck you vsc

GRR

san7890 committed:
documentation amendment

wait that dmdoc was dated

san7890 committed:
forgot some punctuation

and i already pushed that last commit to remote i'm dying walter

Mothblocks commented:
@MrStonedOne Are you ok with this?

Mothblocks commented on code/modules/admin/verbs/config_helpers.dm:
```diff
@@ -0,0 +1,16 @@
+	if(!check_rights(R_SERVER))
+		return
+
+	if(SSjob.generate_config(usr))
```
Can you add an extra input prompt to this, also letting the admin know that if they're not the host, they probably don't want this?

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +603,190 @@ SUBSYSTEM_DEF(job)
+	var/list/file_data = list()
+	config_documentation = initial(config_documentation) // Reset to default juuuuust in case.
+
+	if(fexists(file(toml_file)))
```
Okay, wait, so what is going to happen when a non-host admin presses this button that's easily accessible to them on the live servers?

san7890 replied:
It checks to see if it has a TOML file in the config directory, and if we do, we pull all the values from the existing jobconfig.toml (via `regenerate_job_config()`) and create a new TOML file that we FTP to the requesting client. If existing, all of the values that are in jobconfig.toml can be accessed by any admin by spending some time poking around all of the job datums in SSjob (since that's the end location of the data we pulled from the config-side jobconfig.toml).

Mothblocks replied:
Ok so an admin can't change the default config at all just by clicking buttons right?

san7890 replied:
No.

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +603,190 @@ SUBSYSTEM_DEF(job)
+				)
+
+			if(starting_positions) // Same pattern as above.
+				file_data["[job_key]"] += list(
```
Why `+= list(` instead of just setting the field directly?

`file_data["[job_key]"]["Spawn Positions"] = starting_positions`

san7890 replied:
I was getting some nasty runtimes for some reason last i recall (it's been a second since i've thought, i'll try it again later today

san7890 replied:
Yeah, found what I kept running into:

![image](https://user-images.githubusercontent.com/34697715/196101686-6a464c43-e5c2-47a0-955e-e3a80d90e2af.png)

![image](https://user-images.githubusercontent.com/34697715/196101709-b399f893-c915-4e71-8924-45ff1fd651cb.png)

It works fine if you do it how I have it on current commit (combining lists), so... I'm not really sure what's going on there.

MrStonedOne replied:
bad index on string accesses means the list is null.

Mothblocks replied:
Okay then this is fine, I just didn't know it could be

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +603,190 @@ SUBSYSTEM_DEF(job)
+				)
+			else // If we can't find anything for this variable, then we just throw in the codebase default with it commented out.
+				file_data["[job_key]"] += list(
+					"#Total Positions" = occupation.total_positions,
```
```suggestion
					"# Total Positions" = occupation.total_positions,
```

Very much a nit so feel free not to care but `# name` should be preferred over `#name` for stuff like this because it lets `Ctrl+/` work back and forth, since `Ctrl+/` always adds a space

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +603,190 @@ SUBSYSTEM_DEF(job)
+				)
+			continue
+		else
+			to_chat(user, span_notice("New job [job_name] (using key [job_key]) detected! Adding to jobconfig.toml using default codebase values..."))
```
I thought MSO didn't want you to use default codebase values (and I tend to agree), what was the deal with that?

MrStonedOne replied:
no i wanted them to, i just didn't want it to get perma saved into the generated config

Mothblocks replied:
Oh I see, they're commented out, okay that makes more sense to me

Mothblocks commented on code/modules/jobs/job_types/_job.dm:
```diff
@@ -123,6 +123,10 @@
 
+	/// String key to track any variables we want to tie to this job in config, so we can avoid using the job title. We CAPITALIZE it in order to ensure it's unique and resistant to trivial formatting changes.
+	/// You'll probably break someone's config if you change this, so it's best to not to.
+	var/config_tag = ""
```
Can you add a unit test that verifies every job has a unique, non-blank, non-spaced config tag?

Mothblocks commented on code/controllers/subsystem/job.dm:
```diff
@@ -590,14 +603,190 @@ SUBSYSTEM_DEF(job)
+				continue
+
+			// If the value is commented out, we assume that the server operate did not want to override the codebase default values, so we skip it.
+			var/default_positions = job_config[job_key]["Total Positions"]
```
All of these should be defines to protect against typos.

san7890 committed:
the "easy" stuff first

san7890 committed:
first iteration of unit tests, a few fixes to base code

san7890 committed:
unit test: second iteration - it actually works

san7890 committed:
ALVIN WHO THE FUCK DID BYOND REGEX (third iteration)

okay the regex thing is partially my fault because it was a bit shit ngl

it still sucks though. buncha re-organization of Run()

san7890 committed:
alright so i was just giving myself extra work

fuck this shit

san7890 committed:
final touches

okay, this should be it for now as far as the unit test. that tooooook a whiiileeee

Mothblocks commented on code/modules/unit_tests/verify_config_tags.dm:
```diff
@@ -0,0 +1,33 @@
+	number_of_keys = length(collected_tags)
+	TEST_ASSERT_EQUAL(number_of_keys, number_of_jobs, "Mismatch between the number of joinable occupations: [number_of_jobs] against the number of unique config tags: [number_of_keys]!")
+
+	for(var/iterated_tag in collected_tags)
```
Instead of doing this, check `if (job_tag in collected_tags)` before you add it, that way you can even print the typepath.

san7890 committed:
iteration seven of the unit test

yeah this is definitely better

san7890 commented:
```txt
FAIL: /datum/unit_test/verify_config_tags 24.2s
	REASON #1: Error for /datum/job/head_of_personnel: Job Head of Personnel has a config_tag HEAD OF PERSONNEL with whitespace in it! Please remove the whitespace (use SCREAMING_SNAKE_CASE rules). at code/modules/unit_tests/verify_config_tags.dm:22
	REASON #2: Error for /datum/job/doctor: The config tag MEDICAL_docTOR, for the job Medical Doctor contains lowercase characters. Please change it to SCREAMING_SNAKE_CASE. at code/modules/unit_tests/verify_config_tags.dm:24
	REASON #3: Error for /datum/job/security_officer: The config tag SECURITY_OFFICER is used by multiple jobs! Found as a duplicate on job Security Officer! Please ensure that each job has a unique config tag. at code/modules/unit_tests/verify_config_tags.dm:27
	REASON #4: Expected 34 to be equal to 35. Mismatch between the number of joinable occupations: 35 against the number of unique config tags: 34! at code/modules/unit_tests/verify_config_tags.dm:32
```

cool

san7890 committed:
documentation update, stack traces, new toml

i think that's it as far as what reviews wanted

san7890 committed:
fixes stuff found in testing

weird runtimes (probably my fault), config updates, documentation, etc. very cool

san7890 committed:
actually why did we update the whole ass var every single time

AffectedArc07 commented on config/jobconfig.toml:
```diff
@@ -0,0 +1,227 @@
+## "# Total Positions" should always be changed to "Total Positions", no additional spacing. 
+## Best of luck editing!
+
+[AI]
+"# Playtime Requirements" = 180
+"# Required Account Age" = 30
+"# Spawn Positions" = 1
+"# Total Positions" = 1
```
Keys with spaces 🤮

Mothblocks replied:
just realized the comments on these are wrong because theyre inside the string, lol

AffectedArc07 replied:
Outstanding move

san7890 replied:
i hate to say it but that was intentional because i couldn't figure out how else I could pass the list into TOML Encode - the whole system is reliant on the keys exactly matching and the "#" forcing a mismatch. i wrote about that here in the [October 3rd post](https://github.com/tgstation/tgstation/pull/70199#discussion_r986316574):

>Okay, I think I understood it now. I'm pretty stupid and lazy (mostly, I couldn't think of a better way to do it), so I just purposefully introduced a key mis-match. Every time we generate jobconfig.toml, we will always write to it with commented out keys.

When I wrote it, the hashtag was meant to act as the pseudo-comment instead of an explicit TOML comment-out, because the end goal was just to get a mismatch between the key in the TOML and any expected keys DM-side. There was probably a better way to do it but it wasn't unintentional
